File: Top Interview 150/Java Code/189-rotate-array.java
// Time:O(n)
// Space O(n)
class Solution {
    public void rotate(int[] nums, int k) {
        int n=nums.length;
        List<Integer>v=new ArrayList<>();
        k=k%n;
        for(int i=n-k;i<n;i++) v.add(nums[i]);            
        for(int i=0;i<n-k;i++) v.add(nums[i]);  
        for(int i=0;i<n;i++) nums[i]=v.get(i);
        
    }
}

// Time:O(n)
// Space O(1)
class Solution {
   public static void reverse(int []nums,int l,int h){
       while(l<h){
          int temp=nums[l];
           nums[l]=nums[h];
           nums[h]=temp;
           l++;h--;
       }
   }
    public void rotate(int[] nums, int k) {
        int n=nums.length;
        k=k%n;
       reverse(nums,0,n-1);
       reverse(nums,0,k-1);
       reverse(nums,k,n-1);     
    }
}

File: Top Interview 150/C++ Code/189-rotate-array.cpp
// Time:O(n)
// Space O(n)
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        vector<int>v;int n=nums.size();
        k=k%n;
        for(int i=n-k;i<n;i++) v.push_back(nums[i]);            
        for(int i=0;i<n-k;i++) v.push_back(nums[i]);
        for(int i=0;i<n;i++) nums[i]=v[i];
    }
};

// Time:O(n)
// Space O(1)
class Solution2 {
public:
    void rotate(vector<int>& nums, int k) {
        k=k%nums.size();
       reverse(nums.begin(),nums.end());
       reverse(nums.begin(),nums.begin()+k);
       reverse(nums.begin()+k,nums.end());
    }
};

